lib/astc_encode.cpp

```diff
@@ -0,0 +1,963 @@
+
+    if (This->_protected->_formatSize.flags & KTX_FORMAT_SIZE_PACKED_BIT)
+        return KTX_INVALID_OPERATION;
+
+    // Basic descriptor block begins after the total size field.
+    const uint32_t* BDB = This->pDfd+1;
+
+    uint32_t num_components, component_size;
+    getDFDComponentInfoUnpacked(This->pDfd, &num_components, &component_size);
+
+    if (component_size != 1)
+        return KTX_INVALID_OPERATION; // Can only deal with 8-bit components
+
+    if (This->pData == NULL) {
+        result = ktxTexture2_LoadImageData((ktxTexture2*)This, nullptr, 0);
+
+        if (result != KTX_SUCCESS)
+            return result;
+    }
+
+    ktx_uint32_t threadCount = params->threadCount;
+    if (threadCount < 1)
+        threadCount = 1;
+
+    astcenc_profile profile{ASTCENC_PRF_LDR_SRGB};
+
+    astcenc_swizzle swizzle{ASTCENC_SWZ_R, ASTCENC_SWZ_G, ASTCENC_SWZ_B, ASTCENC_SWZ_A};
+
+    uint32_t        block_size_x{6};
+    uint32_t        block_size_y{6};
+    uint32_t        block_size_z{1};
+    float           quality{ASTCENC_PRE_MEDIUM};
+    uint32_t        flags{params->normalMap ? ASTCENC_FLG_MAP_NORMAL : 0};
+
+    astcBlockSizes(params->blockSize,
+                   block_size_x, block_size_y, block_size_z);
+    quality = astcQuality(params->qualityLevel);
+    profile = astcEncoderAction(*params, BDB);
+    swizzle = astcSwizzle(*params);
+
+    astcenc_config   astc_config;
+    astcenc_context *astc_context;
+    astcenc_error astc_error = astcenc_config_init(profile,
+                                                   block_size_x, block_size_y, block_size_z,
+                                                   quality, flags,
+                                                   &astc_config);
+
+    if (astc_error != ASTCENC_SUCCESS) {
+        std::cout << "Astc config init failed\n";
+        return KTX_INVALID_OPERATION;
+    }
+
+    astc_error  = astcenc_context_alloc(&astc_config, threadCount,
+                                        &astc_context);
+
+    if (astc_error != ASTCENC_SUCCESS) {
+        std::cout << "Astc context alloc failed\n";
+        return KTX_INVALID_OPERATION;
+    }
+
+    //
+    // Copy images into compressor readable format.
+    //
+    std::vector<astcenc_image*> astc_input_images;
+    std::vector<ktx_size_t> astc_image_sizes;
+
+    // Conservative allocation, only used to resize data vectors
+    uint32_t num_images = MAX(1, This->numLayers) * MAX(1, This->numFaces) *
+        MAX(1, This->numLevels) * MAX(1, This->baseDepth);
+
+    astc_input_images.reserve(num_images);
+    astc_image_sizes.reserve(num_images);
+
+    // Size of all astc compressed images in bytes
+    ktx_size_t astc_images_size = 0;
+
+    ktxTexture2_private& priv = *This->_private;
+
+    ktx_size_t level_offset = 0;
+
+    // Walk in reverse on levels so we don't have to do this later
+    // This->numLevels = 0 not allowed for block compressed formats
```

Comment:
If numLevels is 0, this function should change it to 1 and the behavior should be described in the documentation. I suspect ktxTexture2_CompressBasisEx needs the same, if you copied this from there.

Reply:
Fixed in https://github.com/KhronosGroup/KTX-Software/pull/433/commits/c01d05d4cfa0353f848c18051d90ba94862115b5

Reply:
It's not properly fixed. I made a new comment at the place the fix is now needed so I will resolve this conversation.